### ISSUE TITLE
TransportJobAction exception

### ISSUE THREAD

**Issue by bputt** [crate/crate#4885]
**CrateDB version**: 1.0.2

**JVM version**: 1.8.0_92-b14

**OS version / environment description**: Centos 7

**Problem description**:

Trace:
```
[ERROR][transport.netty] failed to handle exception response [io.crate.action.job.TransportJobAction$2@31f05ad1]
java.lang.IndexOutOfBoundsException: Index: 1, Size: 1
at java.util.ArrayList.rangeCheck(ArrayList.java:653)
at java.util.ArrayList.get(ArrayList.java:429)
at io.crate.jobs.PageDownStreamContext.setExhaustedUpstreams(PageDownstreamContext.java:195)
at io.crate.jobs.PageDownStreamContext.failure(PageDownstreamContext.java:150)
at io.crate.executor.transport.executionphases.SetBucketAction.failed(SetBucketAction.java:59)
at io.crate.executor.transport.executionphases.SetBucketActionListener.onFailure(SetBucketActionListener.java:50)
at io.crate.executor.transport.DefaultTransportResponseHandler.handleException(DefaultTramsportResponseHandler.java:62)
at org.elasticsearch.transport.netty.MessageChannelHandler.handleException(MessageChannelHandler.java:184)
at org.elasticsearch.transport.netty.MessageChannelHandler.handleResponse(MessageChannelHandler.java:163)
at org.elasticsearch.transport.netty.MessageChannelHandler.messageReceived(MessageChannelHandler.java:124)
at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:462)
at org.jboss.netty.handler.codec.frame.FrameDecoder.callDecode(FrameDecoder.java:443)
at org.jboss.netty.handler.codec.frame.FrameDecoder.messageReceived(FrameDecoder.java:303)
at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
com.google.common.util.concurrent.ExecutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$6@1eefc5b with executor INSTANCE
java.lang.NullPointerException

```

**Steps to reproduce**: Occasionally occurs when executing multiple batch updates.

Batch updates of 250 at a time
```UPDATE table SET field = ? WHERE _id = ?```

Bulk config settings:

```
threadpool.bulk.type: fixed
threadpool.bulk.size: 2500
threadpool.bulk.queue_size:2000
```

6 Node Cluster (8cpu x 64 gb / node)

Table schema

```
CREATE TABLE table (
    a STRING PRIMARY KEY,
    b LONG PRIMARY KEY,
    c INTEGER,
    d INTEGER,
    e INTEGER,
    f STRING,
    g INTEGER,
    h INTEGER,
    i LONG,
    j INTEGER,
    k INTEGER,
    l LONG,
    m INTEGER,
    n INTEGER,
    o STRING,
    p STRING,
    q STRING,
    r STRING,
    s STRING,
    t STRING,
    u array(object as (
        a STRING,
        b STRING INDEX OFF
    )),
    v array(object as (
        a STRING,
        b STRING,
        c LONG,
        d LONG,
        e INTEGER,
        f INTEGER,
        g INTEGER
    ))
) CLUSTERED INTO 5 SHARDS
WITH (number_of_replicas = 3)
```

**Comment by mfussenegger** [comment]
Thanks for reporting, we'll take a look at it.

**Comment by bputt** [comment]
@mfussenegger another log, not sure if it's related or not

```
com.google.common.util.concurrent.ExcutionList executeListener
SEVERE: RuntimeException while executing runnable com.google.common.util.concurrent.Futures$6@57aa477b with executor INSTANCE
java.lang.NullPointerException
at io.crate.rest.action.RestBulkRowCountReceive.setNextRow(RestBulkRowCountReceiver.java:44)
at io.crate.protocols.postgres.BulkPortal$1.onSuccess(BulkPortal.java:152)
at io.crate.protocols.postgres.BulkPortal$1.onSuccess(BulkPortal.java:141)
at com.google.common.util.concurrent.Futures$6.run(Futures.java:1319)
at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457)
at com.google.common.util.concurrent.ExecutionList.executeListener(ExecutionList.java:156)
at com.google.common.util.concurrent.ExecutionList.execute(ExecutionList.java:145)
at com.google.common.util.concurrent.AbstractFuture.set(AbstractFuture.java:185)
at com.google.common.util.concurrent.Futures$CombinedFuture.setOneValue(Futures.java:1764)
at com.google.common.util.concurrent.Futures$CombinedFuture.access$400(Futures.java:1608)
at com.google.common.util.concurrent.Futures$CombinedFuture$2.run(Futures.java:1686)
at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:457)
at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
```

**Comment by mfussenegger** [comment]
This stacktrace with the NullPointerException is related to https://github.com/crate/crate/issues/4484 for which we've already got a fix.


Regarding the IndexOutOfBoundsException:

So for I haven't been able to reproduce this.
Are you sure that this is caused by the update statements and not something else?
Are you able to reliable reproduce the error on your side?

**Comment by bputt** [comment]
I'm pretty sure it's due to an update statement as the script outputs when an update is finished and it just stalls.

I can replicate the issue when running the script that updates the field. This error doesn't occur on all of the nodes.

The script runs 4 processes that pulls data for a particular day, runs map reduce on one of the columns to get the value for a new column.

We could wait for the release of #4484 to see if this issue still occurs too?

Once one of the processes is finished pulling the data / running map reduce, it'll send bulk updates to the table.

**Comment by mfussenegger** [comment]
> We could wait for the release of #4484 to see if this issue still occurs too? 

Yes, I guess that is an option.

Btw, is there a reason you're using `_id` in the WHERE clause instead of filtering on the defined primary key columns?
If the schema has explicit primary keys the optimizer doesn't pick up `_id` as primary key, so the operations will be less efficient.

I first made my test with `UPDATE table SET field = ? WHERE pkcol1 = ? and pkcol2 = ?` and wasn't able to reproduce https://github.com/crate/crate/issues/4484 because it only happens if you query on a non-primary key (or _id in this case, because there are explicit primary key columns)

So if you changed the query I think the NPE problem would go away and you also wouldn't have to increase the thread_pool queue size.

**Comment by mfussenegger** [comment]
The fix for https://github.com/crate/crate/issues/4484 was released with 1.0.4 - have you been able to update and test if the issue you mentioned here still occurs?

**Comment by joemoe** [comment]
hi @bputt, i'm closing this issue for now. if this still occurs on your cluster, feel free to reopen it any time.
